modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+    */
+    proc writeThis(ch: channel) {
+      _enter();
+
+      //
+      // TODO TODO TODO TODO
+      //
+      // Presently, if I try to let these calls throw, then they will trigger
+      // an error in modules/standard/IO.chpl along the lines of "error: call
+      // to throwing function writeThis without throws, try, or try!", so
+      // I am not sure what the most idiomatic way to write this is.
+      //
+      try {
+        ch.write("[");
+
+        for i in 1..(_size - 1) do
+          ch.write(_getRef(i), ", ");
+
+        if _size > 0 then
+          ch.write(_getRef(_size));
+
+        ch.write("]");
+      } catch e {
+        _leave();
+      }
+
+      _leave();
+    }
+
+    /*
+      Returns `true` if this List contains zero elements.
+
+      :return: `true` if this List is empty.
+      :rtype: `bool`
+    */
+    inline proc isEmpty(): bool {
+      _enter();
+      var result = (_size == 0);
+      _leave();
+      return result;
+    }
+
+    /*
+      The current number of elements contained in this List.
+    */
+    inline proc const size {
+      _enter();
+      var result = _size;
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this List.
+
+      :return: A new DefaultRectangular array.
+    */
+    proc toArray(): [] eltType {
+      _enter();
+
+      //
+      // TODO: Do we want elements to do a full init here, or not?
+      //
+      var result: [1.._size] eltType;
+
+      for i in 1.._size do
+        result[i] = _getRef(i);
+
+      _leave();
+
+      return result;
+    }
+
+  } // End class List!
+
+  /*
+    Clear the contents of this List, then extend this now empty List with the
+    elements contained in another List.
+
+    .. warning::
+
+      Assigning another List to this List will invalidate any references to
+      elements previously contained in this List.
+
+    :arg lhs: The List object to assign to.
+    :arg rhs: The List object to assign from. 
+  */
+  proc =(ref lhs: List(?t, ?), const ref rhs: List(t, ?)) {
+    lhs.clear();
+    lhs.extend(rhs);
+  }
+
+  /*
+    Returns `true` if the contents of two Lists are the same.
+
+    :arg a: A List to compare.
+    :arg b: A List to compare.
+
+    :return: `true` if the contents of two Lists are equal.
+    :rtype: `bool`
+  */
+  proc ==(const ref a: List(?t, ?), const ref b: List(t, ?)): bool {
+    if a.size != b.size then
+      return false;
+
+    for i in 1..(a.size) do
+      if a[i] != b[i] then
+        return false;
+
+    return true;
+  }
+
+  /*
+    Return `true` if the contents of two Lists are not the same.
+
+    :arg a: A List to compare.
+    :arg b: A List to compare.
+
+    :return: `true` if the contents of two Lists are not equal.
+    :rtype: `bool`
+  */
+  proc !=(const ref a: List(?t, ?), const ref b: List(t, ?)): bool {
+    return !(a == b);
+  }
```

Comment:
I think it'd be interesting to have a `compare` method that accepted a comparator, but that is not necessary in the first PR.